captum/optim/_param/image/transforms.py

```diff
@@ -459,3 +466,98 @@ def __init__(self, warp: bool = False) -> None:
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         assert x.dim() == 4
         return nchannels_to_rgb(x, self.warp)
+
+
+class RandomCrop(nn.Module):
+    """
+        Randomly crop out a specific size from an NCHW image tensor.
+    ​
+        Args:
+            crop_size (int, sequence, int): The desired cropped output size.
+    """
+
+    def __init__(
+        self,
+        crop_size: Union[Tuple[int, int], Tuple[int]],
+    ) -> None:
+        super().__init__()
+        self.crop_size = crop_size
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        assert x.dim() == 4
+        hs = x.shape[2] - self.crop_size[0]
+        ws = x.shape[3] - self.crop_size[1]
+        shifts = [
+            torch.randint(low=-hs, high=hs, size=[1]),
+            torch.randint(low=-ws, high=ws, size=[1]),
+        ]
+        x = torch.roll(x, shifts, dims=(2, 3))
+        return center_crop(
+            x,
+            crop_vals=self.crop_size,
+            pixels_from_edges=False,
+        )
+
+
+class AlphaChannelLoss(nn.Module):
```

Comment:
I rewrote AlphaChannelLoss as a loss function now we know it's possible to hook `NaturalImage`, and it seems to work correctly! I called the new objective `BalancedAlphaChannel`, but we may be able to think of a better name.


```
from typing import Optional, Tuple
import torch
import torch.nn as nn


@opt.loss.loss_wrapper
class BalancedAlphaChannel(opt.loss.BaseLoss):
    """
    Self balancing alpha channel loss, that balances opaque and transparent
    regions. Inputs are meant to be NCHW RGBA image tensors with 4 channels.

    See: https://distill.pub/2018/differentiable-parameterizations/
    Mordvintsev, et al., "Differentiable Image Parameterizations", Distill, 2018.
    ​
    Args:
        scale (float, sequence): Tuple of rescaling values to randomly select from.
        crop_size (int, sequence, int, optional): The desired cropped output size
            for secondary alpha channel loss.
        background (tensor, optional):  An NCHW image tensor to be used as the
            alpha channel's background.
    """

    def __init__(
        self,
        target: nn.Module,
        scale,  #: NumSeqOrTensorType,
        crop_size: Tuple[int, int] = None,
        background: Optional[torch.Tensor] = None,
        batch_index: Optional[int] = None,
    ) -> None:
        opt.loss.BaseLoss.__init__(self, target, batch_index)
        self.random_scale = opt.transforms.RandomScale(scale=scale)
        self.random_crop = opt.transforms.RandomCrop(crop_size)
        self.blend_alpha = opt.transforms.BlendAlpha(background=background)

    def __call__(self, targets_to_values):  #: ModuleOutputMapping) -> torch.Tensor:
        activations = targets_to_values[self.target]
        activations = activations[self.batch_index[0] : self.batch_index[1]]
        assert activations.dim() == 4 and activations.size(1) == 4

        shifted = torch.cat(
            [self.blend_alpha(activations.clone()), activations.clone()[:, 3:]], 1
        )

        shifted = self.random_scale(shifted)
        shifted_crop = self.random_crop(shifted)

        return (1.0 - shifted[:, 3:].mean()) + (
            (1.0 - shifted_crop[:, 3:].mean()) * 0.5
        )


def vis_alpha(
    model: torch.nn.Module, target: torch.nn.Module, channel: int
) -> torch.Tensor:
    image = opt.images.NaturalImage((224, 224), channels=4).to(device)
    transforms = torch.nn.Sequential(
        torch.nn.ReflectionPad2d(4),
        opt.transforms.RandomSpatialJitter(8),
        opt.transforms.RandomScale(scale=(1, 0.975, 1.025, 0.95, 1.05)),
        torchvision.transforms.RandomRotation(degrees=(-5, 5)),
        opt.transforms.RandomSpatialJitter(2),
        opt.transforms.CenterCrop((224, 224)),
        opt.transforms.BlendAlpha(),
    )

    loss_fn = opt.loss.NeuronActivation(target, channel_index=channel)
    loss_fn = loss_fn + BalancedAlphaChannel(
        image, scale=[0.6, 0.7, 0.8, 0.9, 1.0, 1.1], crop_size=(160, 160)
    )
    obj = opt.InputOptimization(model, loss_fn, image, transforms)
    history = obj.optimize(opt.optimization.n_steps(128, True))
    image().show()
    image().export("alpha_test.png")


vis_alpha(model, model.mixed4c, 181)
```

![alpha_test](https://user-images.githubusercontent.com/10626398/114733110-4e0d7e00-9d00-11eb-9484-c5dcbae31306.png)


The objective is based off of the Parameterization with Transparency tutorial notebook for Lucid: https://colab.research.google.com/github/tensorflow/lucid/blob/master/notebooks/differentiable-parameterizations/transparency.ipynb

Since the objective includes transforms, maybe we might want to have it in one of the tutorial notebooks instead?